scripts/dev/tidy.php

```diff
@@ -0,0 +1,571 @@
+                     overwrite something by accident.
+
+    -f, --fix        Edit and save files in place
+
+    -q, --quiet      Do not output any message
+
+  EXAMPLES:
+
+    Display help: php tidy.php -h
+
+    Check php-src: scripts/dev/tidy.php .
+
+    Fix php-src: scripts/dev/tidy.php --fix --no-backup .
+
+    Fix custom path or file: ./tidy.php --fix input-path
+
+    Trim only trailing whitespace: ./tidy.php -w --fix input-path
+
+HELP;
+}
+
+/**
+ * Return short invalid usage error.
+ */
+function invalidUsage(string $error): string
+{
+    $output = "tidy.php: $error\n";
+    $output .= "Try 'php tidy.php --help' for more information.\n";
+
+    return $output;
+}
+
+/**
+ * Trim trailing spaces and tabs from each line.
+ */
+function trimTrailingWhitespace(string $content): string
+{
+    return preg_replace('/[\t ]+(\r?$)/m', '$1', $content);
+}
+
+/**
+ * If string doesn't have a final newline append one using a prevailing EOL
+ * character in the current string LF (\n), CRLF (\r\n) or CR (\r).
+ */
+function insertFinalNewline(string $content): string
+{
+    if (
+        substr($content, -2) !== "\r\n"
+        && substr($content, -1) !== "\n"
+        && substr($content, -1) !== "\r"
+    ) {
+        $content .= getEol($content);
+    }
+
+    return $content;
+}
+
+/**
+ * Get all final newlines.
+ */
+function getFinalNewlines(string $content): array
+{
+    $newlines = [];
+
+    while ($content) {
+        if (substr($content, -2) === "\r\n") {
+            $newlines[] = "\r\n";
+            $content = substr($content, 0, -2);
+            continue;
+        }
+
+        if (substr($content, -1) === "\n") {
+            $newlines[] = "\n";
+            $content = substr($content, 0, -1);
+            continue;
+        }
+
+        if (substr($content, -1) === "\r") {
+            $newlines[] = "\r";
+            $content = substr($content, 0, -1);
+            continue;
+        }
+
+        break;
+    }
+
+    return array_reverse($newlines);
+}
+
+/**
+ * Trim final newlines and append an EOL (end of line) based on the previous
+ * prevailing EOL type used LF (\n), CRLF (\r\n) or CR (\r).
+ *
+ * @param int $numOfLines Allowed redundant trailing newlines (default 1).
+ */
+function trimFinalNewlines(string $content, int $numOfLines = 1): string
+{
+    $newlines = getFinalNewlines($content);
+
+    $trimmed = rtrim($content, "\r\n");
+
+    if ($numOfLines === 1) {
+        return $trimmed.getEol($content);
+    }
+
+    if (empty($newlines) && $numOfLines > 0) {
+        return $trimmed.getEol($content);
+    }
+
+    for ($i = 0; $i < $numOfLines; $i++) {
+        if (!empty($newlines[$i])) {
+            $trimmed .= $newlines[$i];
+        }
+    }
+
+    return $trimmed;
+}
+
+/**
+ * Get end of line based on the prevailing LF, CRLF or CR newline characters.
+ */
+function getEol(string $content): ?string
+{
+    // Empty files don't need a newline attached
+    if (empty($content)) {
```

Comment:
Is there any chance to *ever* change `(bool) '0'` to `true`? While this would be an ugly subtle kind of BC break, it seems to me that it might fix more code than break it (I might be totally wrong, though).